### PR TITLE
[ML] add new snapshot upgrader API for upgrading older snapshots

### LINKED ISSUE #64154: [ML] Add functionality to upgrade ML model state to 7.x format before upgrade to 8.0
Currently the `autodetect` process has to know how to load model state formats going right back to version 5.5. In 8.x we would like to drop support for loading model states that predate 7.0. This means that during 7.x we must offer a way for users to easily upgrade model snapshots in 5.x or 6.x formats into the latest 7.x format.

Work was done for elastic/ml-cpp#1460 to add the necessary building block on the C++ side.  What is required now is Java code to make use of this.

The Java work consists of:

* Add a new "upgrade model snapshot" API.
  * To match the other model snapshot APIs this should have job ID and snapshot ID as path arguments.
  * If the referenced snapshot exists and has a `min_version` earlier than 7.0 (or doesn't have a `min_version` field at all) then:
    * The `autodetect` process should be started for the job, and passed the supplied model snapshot (in preference to the one that would normally be restored).
    * A `w` control message should be sent to the `autodetect` process, supplying the same snapshot ID, snapshot timestamp and snapshot description that were on the original snapshot - this will cause it to overwrite the original snapshot  documents with replacement documents in the latest format.
    * The `autodetect` process should be gracefully stopped by closing its input stream - it will not persist state again as no data was sent to it.
* Changes to the deprecations API to report old model snapshots.
  * When the deprecations API is called, it should return a critical problem that must be fixed before upgrade for every model snapshot whose `min_version` is older than 7.0 (or `min_version` not present).

There are many tricky details to work through with the seemingly simple "start and stop `autodetect`" portion of the work.  Will we reuse the same persistent task that we use to open the job for normal operation?  If so, how will we prevent data being sent to the process?  And would that mean the job would have to be closed during the upgrade (not ideal as it could be inconvenient)?  But if we don't reuse the same persistent task then how will we account for memory requirement and enforce that it runs on an ML node?  And how will we avoid named pipe name clashes?  Additionally, when the model snapshot gets persisted for one of these special upgrade invocations of `autodetect`, we need to persist the upgraded model snapshot but _not_ set it as the active one for the job, so the results handling code will need a tweak too.

### PR DESCRIPTION
**Pull request by benwtrent**
This new API provides a way for users to upgrade their own anomaly job
model snapshots.

To upgrade a snapshot the following is done:
- Open a native process given the job id and the desired snapshot id
- load the snapshot to the process
- write the snapshot again from the native task (now updated via the
  native process)

relates #64154

### CODE REVIEW

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java
Comment (benwtrent) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java:2821]: I am not 100% how to do this. Right now this test effectively just checks that the parameters are parsed and sent as the resulting error indicates that we at least tried to load the model snapshot.

The solution might be to get an actual snapshot, and then manually updating the doc so that the min_version is old.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/UpgradeJobModelSnapshotAction.java
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/UpgradeJobModelSnapshotAction.java:35]: This is a master node action as we always want the latest cluster state information.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeState.java
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeState.java:17]: I didn't opt for a "writing_old_state" or an "opened" state as neither really conveyed any information. If the state is null, we know that either it is not assigned to a node or it is assigned and still loading the old snapshot. 

Once we are in the `reading_new_state`, then that indicates that we have reached the point of no return and any failure from that state indicates a corrupted job model snapshot.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:687]: `pipelineId` is for renaming the resulting file pipeline. See below comments for further explanation

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:104]: This just eliminates the small edge cases of requiring job node assignment to take into account node version. These processes are short lived, and restricting the cluster to not be a mixed cluster is a sane limitation. Especially since this API is meant to be used right before upgrading to the next major version.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:120]: this limitation is effectively a throttling limitation. While conceptually it is possible to upgrade numerous snapshots for the same job, we don't want users to do incidentally overload their cluster and prevent other ML jobs from being started.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:168]: I opted for a failure here instead of a silent OK as it seemed nicer to say "hey, we didn't do anything" than that the update was completed.

Especially since this API should only be used when a deprecation warning indicates that a snapshot is too old to use in the next major version.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:181]: It doesn't make sense to upgrade the currently used snapshot of the job. If a job has been closed for an extended period of time (two major versions to be exact), it is probably a good idea to open it anyways for other reasons (job config updates, etc.)

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java:191]: It seemed weird to me that I had to write the header ahead of time as I only wanted to load and then persist the snapshot, but I ran into weird failures when I didn't. 

I can dig further if desired.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java:231]: The regular autodetect communicator doesn't have this issue as the callbacks don't ever feed back into the same executor pool. But here, it does, and to handle it, I am pushing off to the utility thread to execute the shutdown (which needs to use the autodetect executor).

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/NativeAutodetectProcessFactory.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/NativeAutodetectProcessFactory.java:75]: One of the requirements was that this upgrade be possible WHILE the referenced job is running. Consequently, the snapshot upgrade task and the job task COULD be assigned to the same node. If the pipeline ID was not given directly, this would cause a file name conflict.

Admittedly, there is already this "unique pipeline flag" but that is a `long` value. I thought it would be nice to include the snapshot ID directly in the pipeline name. It makes the resulting logs very easy to investigate snapshot upgrader issues by looking for `<job_id>-<snapshot_id>`

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/JobSnapshotUpgraderResultProcessor.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/JobSnapshotUpgraderResultProcessor.java:46]: I created a new processor here as I didn't want to chance ANY other result being written back. This just protects us from inadvertently updating the job results/state when we didn't mean to.

I possibly could have had an `AbstractResultProcessor` class, but shared code was so little, it didn't really seem worth it.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/snapshot/upgrader/SnapshotUpgradePredicate.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/snapshot/upgrader/SnapshotUpgradePredicate.java:56]: In all my testing, the task is only `null` when it has been removed from cluster state. Since this predicate runs AFTER we have confirmed the task has been added to state (the `start task API`), it is good to assume that `null` is removal and thus is completion.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/snapshot/upgrader/SnapshotUpgradeTaskExecutor.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/snapshot/upgrader/SnapshotUpgradeTaskExecutor.java:109]: If we are assigned to a new node while `reading_new_state`, the snapshot could be corrupted since the files are being overwritten one by one.

Consequently, we audit, log and then delete the snapshot as it is unusable anyways.

It MIGHT be better to add a flag to the snapshot that says "bad snapshot". But, the way to recover here would be to delete the job model state and then restore from an elasticsearch snapshot...Up for debate.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/MlJobSnapshotUpgradeIT.java
Comment (benwtrent) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/MlJobSnapshotUpgradeIT.java:146]: After backport I want to add a mixed cluster test (to make sure the mixed node error throws) and I want to verify that the `min_version` is updated on the new snapshot.

Right now, since 8.x does not support upgrades from 6.x, that is not possible here. But in 7.x, it will be good to test that `min_version` gets adjusted.

**Review by benwtrent — COMMENTED** [review]
There was one thing in this change that was concerning to me. 

The model size stats seem to simply disappear when upgrading the snapshot. This is probably because the `CResourceMonitor::createMemoryUsageReport` simply generates the report based on the current job usage (which is effectively nothing). 

This strikes me as strange as `revert` actually does change the model size stats for the job...

Here were my test results:
BEFORE
```
"model_size_stats" : {
        "job_id" : "largish-kibana-sample-data",
        "result_type" : "model_size_stats",
        "model_bytes" : 4972992,
        "peak_model_bytes" : 2495692,
        "model_bytes_exceeded" : 0,
        "model_bytes_memory_limit" : 524288000,
        "total_by_field_count" : 130,
        "total_over_field_count" : 0,
        "total_partition_field_count" : 129,
        "bucket_allocation_failures_count" : 0,
        "memory_status" : "ok",
        "categorized_doc_count" : 0,
        "total_category_count" : 0,
        "frequent_category_count" : 0,
        "rare_category_count" : 0,
        "dead_category_count" : 0,
        "failed_category_count" : 0,
        "categorization_status" : "ok",
        "log_time" : 1604588763336,
        "timestamp" : 1607272200000
      },
```

AFTER
```
"model_size_stats": {
                "job_id": "largish-kibana-sample-data",
                "result_type": "model_size_stats",
                "model_bytes": 11954,
                "peak_model_bytes": 0,
                "model_bytes_exceeded": 0,
                "model_bytes_memory_limit": 524288000,
                "total_by_field_count": 130,
                "total_over_field_count": 0,
                "total_partition_field_count": 129,
                "bucket_allocation_failures_count": 0,
                "memory_status": "ok",
                "categorized_doc_count": 0,
                "total_category_count": 0,
                "frequent_category_count": 0,
                "rare_category_count": 0,
                "dead_category_count": 0,
                "failed_category_count": 0,
                "categorization_status": "ok",
                "log_time": 1604591566986,
                "timestamp": 1607272200000
            },
```

There might need to be changes on the C++ side so that model size states are not regenerated on snapshot save (especially when persist in foreground is used).

@droberts195 let me know what you think.

**Comment by benwtrent** [comment]
After this is merged and backported, another API should be added to check the stats of snapshot upgrades and changes to the deprecation API to include when a snapshot is too old to run in the next major.

**Comment by droberts195** [comment]
> Here were my test results:
>BEFORE
>```
>"model_size_stats" : {
>        "job_id" : "largish-kibana-sample-data",
>        "result_type" : "model_size_stats",
>        "model_bytes" : 4972992,
>        "peak_model_bytes" : 2495692,
>        "model_bytes_exceeded" : 0,
>        "model_bytes_memory_limit" : 524288000,
>        "total_by_field_count" : 130,
>        "total_over_field_count" : 0,
>        "total_partition_field_count" : 129,
>        "bucket_allocation_failures_count" : 0,
>        "memory_status" : "ok",
>        "categorized_doc_count" : 0,
>        "total_category_count" : 0,
>        "frequent_category_count" : 0,
>        "rare_category_count" : 0,
>        "dead_category_count" : 0,
>        "failed_category_count" : 0,
>        "categorization_status" : "ok",
>        "log_time" : 1604588763336,
>        "timestamp" : 1607272200000
>      },
>```
>
> AFTER
>```
>"model_size_stats": {
>                "job_id": "largish-kibana-sample-data",
>                "result_type": "model_size_stats",
>                "model_bytes": 11954,
>                "peak_model_bytes": 0,
>                "model_bytes_exceeded": 0,
>                "model_bytes_memory_limit": 524288000,
>                "total_by_field_count": 130,
>                "total_over_field_count": 0,
>                "total_partition_field_count": 129,
>                "bucket_allocation_failures_count": 0,
>                "memory_status": "ok",
>                "categorized_doc_count": 0,
>                "total_category_count": 0,
>                "frequent_category_count": 0,
>                "rare_category_count": 0,
>                "dead_category_count": 0,
>                "failed_category_count": 0,
>                "categorization_status": "ok",
>                "log_time": 1604591566986,
>                "timestamp": 1607272200000
>            },
>```
>
> There might need to be changes on the C++ side so that model size states are not regenerated on snapshot save (especially when persist in foreground is used).
>
> @droberts195 let me know what you think.

I agree there need to be changes on the C++ side.

For `model_bytes`, I think the `w` control message should trigger a full recalculation of model memory before persisting state, i.e. iterate all the resources that `model::CResourceMonitor` knows about and update their memory usage.  This will result in an accurate figure for the latest data structures.

For `peak_model_bytes` it's slightly worrying that it gets reset to zero, as it's supposed to be taken from a persistent counter.  It suggests we might not be restoring the values of persistent counters correctly.  If so this is a bug that needs to be fixed soon.

@edsavage please could you investigate those two things.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java
Comment (droberts195) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java:2821]: > The solution might be to get an actual snapshot, and then manually updating the doc so that the min_version is old.

Yes, in terms of testing the infrastructure that would be a good way.  Run a simple job like farequote, update the model snapshot document after the job is closed to have a `min_version` from the previous major, then upgrade it.  Not sure this needs to be done in the HRLC tests though - for such a complex test the native multi node tests seems like the single place to do it.  I am happy to leave this test as-is, just testing the parameter passing.

In terms of testing actual upgrade, it could be done in the BWC tests.  We could have a BWC test (Java, not YAML) that does nothing when the old cluster is on the same major, but when the old cluster is on a different major it opens/runs/closes a job in the old cluster, then upgrades its model snapshot in the fully upgraded cluster (and does nothing in the mixed cluster).

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/MlClientDocumentationIT.java
Comment (droberts195) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/MlClientDocumentationIT.java:2377]: As above, this will be complex and expensive, and I am not sure that is justified for the docs tests.  We can do it once as part of the native multi node tests, but burning that CPU many times in a full CI run seems unjustified.

File: docs/java-rest/high-level/ml/upgrade-job-model-snapshot.asciidoc
Comment (droberts195) [inline docs/java-rest/high-level/ml/upgrade-job-model-snapshot.asciidoc:27]: As an end user I would be interested in what this means if `wait_for_completion=false`, and what the default is.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeState.java
Comment (droberts195) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeState.java:17]: `reading_new_state` is very much from the perspective of the Java code rather than the end user.  As an end user who doesn't even know that the code is split between Java and C++ I would have thought `writing_new_state` makes more sense.  Or `saving_new_state` would be a compromise that makes sense to both end users and Java developers.

I would also introduce a `reading_old_state` or `loading_old_state` enum value that can be used in stats and API responses instead of `null`.  We went through that cycle with job states.  Initially there was no `opening` state, because a `null` task state basically meant that.  But then we found it was nicer to have a specific enum value for it and translate `null` to that enum value in some places.  Even if it's not used anywhere it will avoid BWC code to add it to the enum from the outset.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeTaskState.java
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeTaskState.java:49]: Did you consider `UncheckedIOException`?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:168]: I would do the check on the exact version rather than just the major.  Although it shouldn't be necessary, upgrading the format from e.g. 7.0 format to 7.11 format might be a useful piece of functionality to have in the future to work around some other bug.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:181]: This means extra complication for the Kibana upgrade assistant though.  For every model snapshot that exists that is too old it will now have to recommend one of two possible courses of action, depending on whether the snapshot is the active one or not.  Opening and closing a job normally without sending it any data doesn't rewrite the snapshot, so the user would also have to feed some data to actually change the active model snapshot of the job.  So I think this check should be altered to only ban upgrading the active snapshot if the job is open.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java
Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java:191]: I am not surprised you had to write a header.  You could probably get away with writing one with just the control field (field name `.`).  But it's not particularly important, so I'm happy to leave what's here.

**Review by droberts195 — COMMENTED** [review]
Thanks for writing this extremely complicated yet also tedious functionality.

There's a lot of code so I haven't reviewed it all in detail, but have left an initial set of comments.  The biggest one is that we need to think about how to avoid excessive complexity in the Kibana migration assistant that will have to use this code eventually.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:168]: I can do that. It just didn't seem 100% necessary given the current state of things. Also, it doesn't seem that the `min_version` is really ever changed to the latest version. So, it very well could be that somebody calls this api numerous times in a row and since `min_version` is never changed, no error is returned. 

I suppose this is "buyer beware" and the API won't hurt anything directly.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java:181]: > So I think this check should be altered to only ban upgrading the active snapshot if the job is open.

100%, this seems good

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java
Comment (benwtrent) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java:2821]: yeah, I have a BWC test class covering this case.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/MlClientDocumentationIT.java
Comment (benwtrent) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/MlClientDocumentationIT.java:2377]: yeah, I have a BWC test class covering this case.

**Review by droberts195 — APPROVED** [review]
LGTM

I noticed a few more minor things but am happy to merge this once they're resolved.

**Pull request merged**
